worker/controlsocket/handlers_test.go

```diff
@@ -0,0 +1,444 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package controlsocket
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"github.com/juju/names/v4"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/core/permission"
+	"github.com/juju/juju/state"
+)
+
+type handlerSuite struct {
+	state  *fakeState
+	logger Logger
+}
+
+var _ = gc.Suite(&handlerSuite{})
+
+type handlerTest struct {
+	// Request
+	method   string
+	endpoint string
+	body     string
+	// Response
+	statusCode int
+	response   string // response body
+}
+
+func (s *handlerSuite) SetUpTest(c *gc.C) {
+	s.state = &fakeState{}
+	s.logger = loggo.GetLogger(c.TestName())
+}
+
+func (s *handlerSuite) runHandlerTest(c *gc.C, test handlerTest) {
+	socket, err := os.CreateTemp("", "")
+	c.Assert(err, jc.ErrorIsNil)
+
+	_, err = NewWorker(Config{
+		State:      s.state,
+		Logger:     s.logger,
+		SocketName: socket.Name(),
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	serverURL := "http://localhost:8080"
+	req, err := http.NewRequest(
+		test.method,
+		serverURL+test.endpoint,
+		strings.NewReader(test.body),
+	)
+	c.Assert(err, jc.ErrorIsNil)
+
+	resp, err := client(socket.Name()).Do(req)
```

Comment:
Use `c.MkDir()` to create a directory for this test run, so the test runner will clean it up.
We are also currently leaking an open fd with `os.CreateTemp` as it returns an open file.

```suggestion
	tmpDir := c.MkDir()
	socket := path.Join(tmpDir, "test.socket")

	_, err = NewWorker(Config{
		State:      s.state,
		Logger:     s.logger,
		SocketName: socket,
	})
	c.Assert(err, jc.ErrorIsNil)

	serverURL := "http://localhost:8080"
	req, err := http.NewRequest(
		test.method,
		serverURL+test.endpoint,
		strings.NewReader(test.body),
	)
	c.Assert(err, jc.ErrorIsNil)

	resp, err := client(socket).Do(req)
```